### ISSUE TITLE
C# tutorials - LINQ

### ISSUE THREAD

**Issue by BethMassi** [dotnet/docs#90]
We need LINQ in the tutorials list. This is a big topic so there could be a lot of samples here. Good starting point: https://code.msdn.microsoft.com/101-LINQ-Samples-3fb9811b 

Tutorials should include one or more code examples, each example should also show expected output and should include an explanation of the code where it isn't obvious. Examples shouldn't be dependent on a specific IDE or target platform.

[BillWagner]  Adding sub-tasks:
- [ ] Bring forward the LINQ overview from the .NET Primer.
- [ ] Provide an overview of LINQ and databases, with extensive links to the EF documentation
- [ ] Provide a reference to Azure Storage LINQ providers.
- [ ] Bring forward explanations from the 101 LINQ samples (code has been added, descriptions have not).

**Comment by matthewreily** [comment]
I'd like to help out on this one and will gladly divide an conquer with anyone else.

**Comment by BethMassi** [comment]
@matthewreily you rock! 
@blackdwarf @cartermp how do you envision the C# tutorials land in the repo folder structure we have? Should we pile into docs/tutorials-samples/CSharp/LINQ ?

**Comment by blackdwarf** [comment]
Are these going to be done in restructured text or markdown?

**Comment by BethMassi** [comment]
I thought they had to be rst. That's what it says in the contributing guide :smile:

**Comment by blackdwarf** [comment]
Ah, ok. So this is interesting. let's keep them where you suggested. Its the thing that makes most sense. The actual code for the tutorials I think should also be there.

**Comment by cartermp** [comment]
The [existing LINQ article](https://github.com/dotnet/core-docs/blob/master/docs/concepts/linq.rst) as some examples which could be shared/altered/used as a starting point for this.

What I'm hoping eventually happens is for any code snippets (be them in tutorials or elsewhere) are embedded directly from some code which sits in the repository, and is buildable.  It would be a good idea to focus on the code snippets in that way now so that we don't have to retroactively go back to every document which uses LINQ and embed buildable code.

**Comment by blackdwarf** [comment]
Sure, rest has that option. It is more of where the said source code will sit, which repo? I suggest this one, for now, that is, co-located with the documents.

**Comment by BethMassi** [comment]
Cool. I think we should start with a quick overview page (pulling from the existing doc) and then further break the content down into folders based on the type of operation. @cartermp @blackdwarf is this how you want the samples, in its own folder next to the doc itself? How much should we spit down to? Is one sample per doc or one doc can have multiple samples?

docs/tutorials-samples/CSharp
/LINQ
  overview.rst
  /restriction-operators
     /where-simple1
       where-simple1.rst
       /sample         
     /where-simple2
     ...
  /projection-operators
  /partitioning-operators
  /ordering-operators
  /grouping-operators
  /set-operators
  /conversion-operators
  /element-operators
  /generation-operators
  /quantifiers
  /aggregate-operators
  /join-operators

**Comment by cartermp** [comment]
I actually like the idea of having a totally separate `/samples` structure that samples are pulled from.  Co-location with docs could get tricky when docs from multiple, unrelated areas pull in example code from the same sample source.

@blackdwarf  what are your thoughts on this?

**Comment by BethMassi** [comment]
Okay I like that proposal - we create a top level /samples folder next to the /docs and use a similar taxonomy for the folder structure under that. What's confusing is we have /tutorials-samples under docs right now. Should we rename to just /tutorials ?

How about this? @cartermp @blackdwarf @matthewreily 

    /docs
	    /languages
	        /csharp
			    /linq
				    overview.rst
				    restriction-operators.rst
                    projection-operators.rst
                    partitioning-operators.rst
                    ordering-operators.rst
                    grouping-operators.rst
                    set-operators.rst
                    conversion-operators.rst
                    element-operators.rst
                    generation-operators.rst
                    quantifiers.rst
                    aggregate-operators.rst
                    join-operators.rst
    /samples
	    /languages
		    /csharp
			    /linq
                    /restriction-operators
					    where-sample1.cs
						where-sample2.cs
						where-sample3.cs
                    /projection-operators
					    ...
					/partitioning-operators
					/ordering-operators
					/grouping-operators
					/set-operators
					/conversion-operators
					/element-operators
					/generation-operators
					/quantifiers
					/aggregate-operators
					/join-operators

**Comment by cartermp** [comment]
I agree that we should rename `/tutorials-samples` to `tutorials` under the docs folder.  Let's go with that.

I think LINQ should be in its own area in the `/samples` directory because it's not tied to C#.  Perhaps something like this would work well:

```
/samples
    /linq
        /csharp
            /restriction-operators
        /vb
            /restriction-operators
```

**Comment by blackdwarf** [comment]
So...how will this work for your proposal on building the samples with the actual code it is...sampling, so to speak @cartermp ?  I'm fine with the overall approach. Renaming tutorials-samples to tutorials also makes sense.

**Comment by cartermp** [comment]
These samples would be built when a corefx library is published.  So a hook would run a script which scans this `/samples` folder, restoring packages and rebuilding the projects.  There would most likely be some sort of indicator in the README listing any samples which are broken and need to be updated.

At least that's what I'm thinking.  This is assuming none of the samples rely on .NET Framework.

**Comment by BethMassi** [comment]
@cartermp WRT the samples structure, I think it would be helpful to have the same (or very close) taxonomy to the docs. So should it be:

    /docs
        /linq
            overview.rst
            /csharp
                restriction-operators.rst
                ...
           /vb
                restriction-operators.rst
                ...
This means that VB & C# code would be separate in each of the tutorials. I think that's good, but we effectively double the number of docs.

**Comment by cartermp** [comment]
The linq tutorial could be written in a way that had side-by-side C# and VB samples.  Something like this:

```
Filtering a list is easy!  Here's how you can do it.

C#:
var filtered = items.Where(item => condition(item));

VB:
Var filtered = items.Where(Function(item) condition(item))
```

The C# line of code would be embedded from the `/csharp` folder in `/samples`, and the VB line would be embedded from the `/vb` folder.  At least that's what I'm thinking.

**Comment by BethMassi** [comment]
But then that means when the community writes docs, they'll have to be bi-lingual. Is that going to be a requirement?

**Comment by mairaw** [comment]
So, are you guys assuming that it'll be a 1:1 relationship to the docs? Meaning there's no reuse of sample between ref and conceptual, for example.
I don't think it should be a requirement but we can help provide the missing language, right?

**Comment by BethMassi** [comment]
I think we need to think about the consumer of the core docs. Will they be bilingual? I don't think so IMO. Will it distract from the tutorial to have multiple languages in the same doc? MSDN lets you pick a preference, I don't think we can do that here, right? I think it's OK to separate them. In @cartermp 's example, I'd rather show the lambda syntax as well as the query expression syntax, but keep it the same language. This way you can elaborate on each language's merits in each doc. I do realize though that this could create a lot of doc files. Thoughts?

**Comment by mairaw** [comment]
Even if we highlight things that are specific for the language, there would still be common content between the languages and that could become an issue to maintain them in sync. Have we discussed that kind of language filtering requirement with Peter's team?

**Comment by BethMassi** [comment]
I think another big question here is what will contributors prefer? If we impose a bilingual format, will people be able to easily contribute to that? If we separate, I could see that being easier in a way. BTW I am a HUGE VB fan particularly WRT LINQ. The syntax there is so eloquent IMO. It definitely starts to diverge in a major way with XML query for instance (in a very good, productive way). I'm just wondering if it's going to be hard to learn something from two different perspectives. I think we need to find a balance, or prioritize on C# first.

**Comment by matthewreily** [comment]
As a consumer of the documentation, I would prefer to have code samples filtered to my language of choice.  Unless I'm trying to compare and contrast syntax between the languages, the other language would just be noise IMO.  Excuse my ignorance in the documentation technology, but Is there a way to toggle the preference somehow or collapse a section? for example something like a #region in a .cs file.

As a contributor, I like the idea of having the conceptual documentation as one document and the corresponding references in their own separate language specific folders.  If I need help with a particular language's reference examples, it wouldn't prevent me from writing the conceptual document and the other language references.  I could then open an issue asking for help with the other language.

**Comment by BethMassi** [comment]
I agree. I'm happy to help with the VB samples with you. :) If we can filter that'd be awesome. @blackdwarf is there something simple we can do to filter languages on a topic?

**Comment by cartermp** [comment]
@BethMassi @matthewreily If we go a route of having `{ one doc, many languages }` for a given document, I think it's alright to have eventual consistency w.r.t the relevant code snippets being in all the supported languages.

**Comment by BethMassi** [comment]
I agree with that but I don't want to have to look at them all.

**Comment by blackdwarf** [comment]
Well, this went off the rails quickly. ;)

There is a solution to this. We can keep the content that is general as the main document, and then just include it in the C#/VB specific pages. reST allows you to include other files, so this means that the pieces of content that apply cross-language are in one place, while specifics (like syntax) are separate. I would like to see the TOC structure for this as well. RTD/Sphinx doesn't have a readily available feature for switching languages *in situ* on the page a-la MSDN, so we would have to be  **very** creative to make this work, and I just don't think that this is a good investment of our time at this point. @mairaw we haven't really discussed this with Peter, I will start the discussion with them. 

Mixing VB and C# (and others) in the same page without some control to remove one or the other seems like wasting people's time to me (no offense). I'm interested, as a consumer, on getting the syntax I want to adopt, and now you're making these jarring drops between two completely different syntaxes on the same page. For the love of God man, why?!?! 

I suggest we focus on the C# tutorial and create good content for it. We already have the primer which goes through the very basic theory behind these and other concepts. With that, we could say that the tutorial is a place to see *how* rather than *what* very quickly, thus making it a problem-solution oriented content for a given language.

**Comment by BethMassi** [comment]
I agree. I want to see one syntax on a page. That's why I asked if it was simple to hide a language on the same page. I guess it's not. So let's focus on C#. And then a separate page later for other languages as they come online.

**Comment by cartermp** [comment]
Sounds fair.  If it's too big of a technical challenge to have filtered languages, we shouldn't have duplicate sample code.

**Comment by matthewreily** [comment]
I'd like to get started on this but am confused with the folder layout we have decide on.  Are we wanting something like the following?

+   docs
      +   linq
         + overview.rst
         + restriction-operators.rst
         + projection-operators.rst
         + partitioning-operators.rst

...
+ samples
   + linq
      + restriction-operators
         + where-sample1.cs
         + where-sample2.cs
         + where-sample3.cs
      + projection-operators
            + ...
      + partitioning-operators
            + ...

**Comment by matthewreily** [comment]
Ready to start working on a some tutorial content.  Should I start with a linq overview doc or a document covering one of the linq samples we created? cc @BethMassi

**Comment by blackdwarf** [comment]
@matthewreily there is already a LINQ overview doc in the .NET Primer, so you can either link to that or you can actually reword it to be better.

**Comment by cartermp** [comment]
Yep!  If there's a way to repurpose, modify, or otherwise rewrite the LINQ overview doc then that would be a great place to start.

**Comment by BillWagner** [comment]
There's a lot of history here, and this is a big topic.

I added sub-tasks in the original description to provide some pointers on where to start.

**Comment by BillWagner** [comment]
@stevehoag Will this issue be handled by what you're doing for #490?  If so, do you want to assign it to yourself and we'll close it when we get that PR

**Comment by sputier** [comment]
@mairaw @BillWagner What is the current state of this issue ? Regarding the previous discussion on this topic, is there a preferred way of organizing and working on it ? 

Is this king of layout (suggested a long time ago) OK ?

    docs
        linq
            overview.rst
            restriction-operators.rst
            projection-operators.rst
            partitioning-operators.rst
    ...

    samples
        linq
            restriction-operators
                where-sample1.cs
                where-sample2.cs
                where-sample3.cs
         
                //Would also where-sample1.vb be located here ?
            projection-operators
            ...
            partitioning-operators
        ...

I am ready to handle this one !

**Comment by mairaw** [comment]
HI @sputier. Thanks for your interest in contributing! I'll let @BillWagner to update you in the status of this since I've merely changed milestones and owners during our triage meetings.

**Comment by BillWagner** [comment]
@sputier We'll want to change the organization for this somewhat.

This should go under samples, and look something like this (using headers for outlining):

# Samples
## LINQ
### Searches with Where
### Projections with Select
### Partitioning with Group by

And so on.

What I'd like to see is the titles point to a problem or use (search, project, partition, filter, etc.) 

For simple tasks, we may want to experiment with our C# interactive experience. You can see an example on [this page](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/foreach-in) 

Also, we should cross-reference to the [LINQ material](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/) in the programming guide as well.

Also, this may be large enough for more than one PR.  Let me know your thoughts.  

Once again, thanks for contributing.

**Comment by sputier** [comment]
@BillWagner Thanks for these informations. Your suggestion about the global structure seems great to me. 

The [C# Guide \ LINQ section](https://docs.microsoft.com/en-us/dotnet/csharp/linq/) already seems to describe different use cases. Regarding this fact, what is the value we want to give to the readers in this section ? A detailed exploration of each LINQ operator ?  A real step-by-step to achieve specific results ? Something other ?

As you said, this will probably be too large for a single PR. Maybe 1 PR per 2/3 sections. This could be defined when I'll have a more precise idea of the total work for this issue, if you're OK with that.

**Comment by BillWagner** [comment]
@sputier Great questions. Let me think a bit, and respond with more guidance in a day or two. I like your thoughts so far.

**Comment by sputier** [comment]
Hi @BillWagner, have you any news about this issue ? I you have a few ideas, i'd be glad to help you elaborate on it !

**Comment by BillWagner** [comment]
Thanks for the reminder @sputier 

I think what will be most useful is a series of short, interactive code snippets that enable readers to see the correct syntax for different kinds of LINQ queries, and explore what the results are.

A simplified version of the "101 samples" that shows the structure of each of the queries would be very helpful. Keeping it small, so types of queries can be found quickly is a good idea.

The [LINQ in C# section](https://docs.microsoft.com/dotnet/csharp/linq/linq-in-csharp) has a lot of what we need. A better overview, and more interactive samples would help. 

You can see an example of how to use the C# interactive REPL in this article: https://github.com/dotnet/docs/blob/master/docs/csharp/language-reference/keywords/foreach-in.md

The key to remember is that your code gets inserted into a Main() method, so you are limited to code that you can include in a single method.

Ping with questions, or open a 'WIP' PR with some outlines and thoughts.

**Comment by sputier** [comment]
@BillWagner The end of the year was a very busy time, so I've set this issue aside for a while, but I'm now back to tackle it. 

Before deep diving into the subject : I'm not sure about the way I should integrate everything in the repository tree... Should I add the code samples under /samples and the text content under docs/samples-and-tutorials ? The file [/docs/samples-and-tutorials/index.md](/dotnet/docs/blob/master/docs/samples-and-tutorials/index.md) already points to [Working with LINQ](/dotnet/docs/blob/master/docs/csharp/tutorials/working-with-linq.md), so I'm a bit confused...

I'll open a WIP PR soon, as I'm working on a first draft of the code samples that will be used in the interactive REPL.

**Comment by BillWagner** [comment]
@sputier You are seeing a lot of the artifacts of the migration of existing MSDN content. We will keep migrating toward a new structure.

For new samples, put complete programs under /samples/csharp. You can make a new directory there for anything you are building.  That way, our CI build will find the .csproj files, and build and run any unit tests.

The MD files should be under docs/csharp/tutorials.  If there are a lot of topics, you can create a subfolder with the different topics. The only requirement for a new folder is that it needs an 'index.md' which is the home page for that folder.

Please tag me when you open a WIP PR, and I'll have a look.

**Comment by rpetrusha** [comment]
The LINQ tutorial is [Working with LINQ](https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/working-with-linq). Closing this issue.
